gemfire-core/src/main/java/com/gemstone/gemfire/distributed/internal/membership/jgroup/JGroupMembershipManager.java

```diff
@@ -2151,12 +2178,17 @@ protected void processMessage(DistributionMessage msg) {
 
         // If it's a new sender, wait our turn, generate the event
         if (isNew) {
+          latestViewLock.readLock().unlock(); // synchronized
           shunned = !addSurpriseMember(m, getStubForMember(m));
         } // isNew
       }
 
       // Latch the view before we unlock
-    } // synchronized
+    } finally {
+      if (!isNew) { // note: if(isNew) then we release read and acquire writeLock inside addSurpriseMember.
+        latestViewLock.readLock().unlock(); // synchronized
+      }
```

Comment:
To be absolutely sure, it will probably be better to set a flag after unlock() above instead of relying on isNew:

```
          // If it's a new sender, wait our turn, generate the event
         if (isNew) {
+          latestViewLock.readLock().unlock(); // synchronized
+          viewReadLocked = false;
           shunned = !addSurpriseMember(m, getStubForMember(m));
         } // isNew
       }
 
       // Latch the view before we unlock
-    } // synchronized
+    } finally {
+      if (viewReadLocked) { // note: if(!viewReadLocked) then we release read and acquire writeLock inside addSurpriseMember.
+        latestViewLock.readLock().unlock(); // synchronized
+      } 
```